### PR TITLE
Addition of Migrator Script

### PR DESCRIPTION
**Pull request by chases2**
Migrates Dashboard Tabs from config.yaml to a given prow directory
Meant to be run as part of a cleanup process for issue #13057

/cc @Katharine

### CODE REVIEW

**Comment by chases2** [comment]
Note: an example of this script's effects can be found in #13205

**Comment by chases2** [comment]
"/io_bazel_rules_go/BUILD.bazel:8:1: ... UNAVAILABLE: 502:Bad Gateway"
Might be network-related flakiness?
/retest

File: experiment/migrator/migrator.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python2
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: nit: please use python3 (ref: https://github.com/kubernetes/test-infra/issues/13164)

File: experiment/migrator/migrator.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+Only moves Dashboard Tabs
+Skips any Dashboard Tab Configuration that contains unusual keys, even if they're incorrect keys.
+"""
+
+import re
+from os import walk
+import ruamel.yaml
+
+PROW_DIR = "../../config/jobs"
+TESTGRID_CONFIG = "../../testgrid/config.yaml"
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: nit: consider passing these in as args, eg: https://github.com/kubernetes/test-infra/blob/eb8efa7a9dcdf20840da42a9944bfbdea99c8600/experiment/fix_testgrid_config.py#L90-L99

File: experiment/migrator/migrator.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                if assertTabKeys(dashTab):
+                    moveTab(dashboard, dashTab, dashNameA)
+
+    confFile.seek(0)
+    ruamel.yaml.dump(config, confFile, Dumper=ruamel.yaml.RoundTripDumper)
+    confFile.truncate()
+    confFile.close()
+
+
+def moveTab(dashboard, dashTab, dashboardName):
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: nit: prefer snake_case for python, eg: `move_tab`

Comment (spiffxp) [reply experiment/migrator/migrator.py]: is `dashboardName` needed? we could get it from `dashboard["name"]`

Comment (spiffxp) [reply experiment/migrator/migrator.py]: meta nit: mixing IO and logic in here makes this difficult to test in isolation, consider extracting reading and writing of prowjob config from this, such that this function just becomes logic

File: experiment/migrator/migrator.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import re
+from os import walk
+import ruamel.yaml
+
+PROW_DIR = "../../config/jobs"
+TESTGRID_CONFIG = "../../testgrid/config.yaml"
+
+def main():
+    confFile = open(TESTGRID_CONFIG, "r+")
+    config = ruamel.yaml.load(confFile, Loader=ruamel.yaml.RoundTripLoader, preserve_quotes=True)
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: consider using `with`, eg: https://github.com/kubernetes/test-infra/blob/eb8efa7a9dcdf20840da42a9944bfbdea99c8600/experiment/fix_testgrid_config.py#L73-L74

File: experiment/migrator/migrator.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        dashNameA = dashboard["name"]
+        if "dashboard_tab" in dashboard:
+            for dashTab in dashboard["dashboard_tab"][:]:
+                if assertTabKeys(dashTab):
+                    moveTab(dashboard, dashTab, dashNameA)
+
+    confFile.seek(0)
+    ruamel.yaml.dump(config, confFile, Dumper=ruamel.yaml.RoundTripDumper)
+    confFile.truncate()
+    confFile.close()
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: consider using `with`, eg: https://github.com/kubernetes/test-infra/blob/eb8efa7a9dcdf20840da42a9944bfbdea99c8600/experiment/fix_testgrid_config.py#L84-L88

File: experiment/migrator/migrator.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    It also deletes from the config the dashboard tab, so there are no duplicates.
+
+    It has a hardcoded exception for fejta-bot-periodics, because bugs
+    """
+    prowJobName = dashTab["test_group_name"]
+    prowJobFileName = findProwJob(prowJobName, PROW_DIR)
+    # Ruamel won't be able to successfully dump fejta-bot-periodics
+    # See https://bitbucket.org/ruamel/yaml/issues/258/applying-json-patch-breaks-comment
+    if (prowJobFileName != ""
+            and prowJobFileName != (PROW_DIR + "/kubernetes/test-infra/fejta-bot-periodics.yaml")):
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: consider returning on the not of this, such that we can avoid nesting everything else below under this if

File: experiment/migrator/migrator.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+def patchProwJob(prowYaml, dashTab, dashboardName):
+    """Given a prow yaml and a compatible dashTab,
+    returns the new prow yaml with a new or amended annotation
+    """
+    if "annotations" in prowYaml:
+        #There exists an annotation; amend it
+        annotation = prowYaml["annotations"]
+        if "testgrid-dashboards" in prowYaml["annotations"]:
+            #Existing annotation is a testgrid annotation
+            #The dashboard name must come first if it's a sig-release-master-* dashboard
```
Comment (spiffxp) [inline experiment/migrator/migrator.py]: out of curiosity, why is this?

Comment (chases2) [reply experiment/migrator/migrator.py]: The email annotation applies the email to "...the tab created in the first dashboard specified in testgrid-dashboards" ([config.md](https://github.com/kubernetes/test-infra/blob/master/testgrid/config.md)), and not all dashboards.

Not having an email directly associated with a sig-release-master-* tab is a blocking failure in the config-checking tests.

**Comment by chases2** [comment]
@spiffxp I've addressed your review and made sure that it acts sanely in Python3 (example output is in #13205), PTAL

File: experiment/migrate_testgrid_tabs.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    if "postsubmits" in prowConfig:
+        for repo, jobs in prowConfig["postsubmits"].items():
+            for (i, job) in enumerate(jobs):
+                if prowJobName == job["name"]:
+                    prowConfig["postsubmits"][repo][i] = patch_prow_job(job,
+                                                                        dashTab,
+                                                                        dashboardName)
+    # periodics -> [{name: prowjob}]
+    if "periodics" in prowConfig:
+        for i in range(0, len(prowConfig["periodics"])):
```
Comment (michelle192837) [inline experiment/migrate_testgrid_tabs.py]: Loop over prowConfig["periodics"].items()?

File: experiment/migrate_testgrid_tabs.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+    # Matching file found; patch and delete
+    print("Patching {0} in {1}".format(prowJobName, prowJobFileName))
+
+    with open(prowJobFileName, "r") as prowJobFile:
+        prowConfig = ruamel.yaml.load(prowJobFile,
+                                      Loader=ruamel.yaml.RoundTripLoader,
+                                      preserve_quotes=True)
+
+    # For each presubmits, postsubmits, periodic:
```
Comment (michelle192837) [inline experiment/migrate_testgrid_tabs.py]: Argh the logic isn't quite the same for all of them. Could combine presubmits and postsubmits logic at least?

File: experiment/migrate_testgrid_tabs.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    # delete tab
+    dashboard["dashboard_tab"].remove(dashTab)
+
+
+def assert_tab_keys(tab):
+    """Assert that the dashboard tab doesn't contain anything not on the
+    following whitelist:
+    name, description, test_group_name, alert_options->alert_mail_to_addresses
+    """
+    failingKeys = [st for st in tab.keys() if st != "name"
```
Comment (michelle192837) [inline experiment/migrate_testgrid_tabs.py]: List comprehensions should probably be a one-liner. I think you can also shorten to:

whitelist = ["name", "description", "test_group_name", "alert_options"]
if any([st for st in tab.keys() if st not in whitelist]):
  return False

File: experiment/migrate_testgrid_tabs.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            if filename.endswith(".yaml") and filename not in PROW_BLACKLIST:
+                for _, line in enumerate(open(dirpath + "/" + filename)):
+                    for _ in re.finditer(pattern, line):
+                        #print "Found %s in %s" % (name, filename)
+                        return dirpath + "/" + filename
+    return ""
+
+
+def patch_prow_job(prowYaml, dashTab, dashboardName):
+    """Given a prow yaml and a compatible dashTab,
```
Comment (michelle192837) [inline experiment/migrate_testgrid_tabs.py]: Summary line, then more detail? https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: experiment/migrate_testgrid_tabs.py (hunk 8 above)
Comment (chases2) [reply experiment/migrate_testgrid_tabs.py]: prowConfig["periodics"] is a list, not a map, so I can just reference `jobs in prowConfig["periodics"]`

File: experiment/migrate_testgrid_tabs.py (hunk 9 above)
Comment (chases2) [reply experiment/migrate_testgrid_tabs.py]: I'm not sure that combining these two will increase readability, but cleaning up the loops with the suggestions you've made certainly have

**Comment by chases2** [comment]
@michelle192837 Nits have been resolved, PTAL

File: experiment/migrate_testgrid_tabs.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    To be migratable, the annotations must contain only the following:
+    name, description, test_group_name, alert_options->alert_mail_to_addresses
+    """
+    whitelist = ["name", "description", "test_group_name", "alert_options"]
+    if [k for k in tab.keys() if k not in whitelist]:
+        return False
+
+    if "alert_options" in tab:
+        alertKeys = tab["alert_options"].keys()
+        if len(alertKeys) != 1 or "alert_mail_to_addresses" not in alertKeys:
```
Comment (spiffxp) [inline experiment/migrate_testgrid_tabs.py]: nit: could redo this as another `whitelist`-style check so it's easier to add a new key if we find we want to support more `alert_options` keys

File: experiment/migrate_testgrid_tabs.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+def find_prow_job(name, path):
+    """Finds a Prow Job in a given subdirectory.
+
+    Returns the first file that contains the named prow job
+    Returns "" if there isn't one
+    Dives into subdirectories
+    Ignores PROW_BLACKLIST
+    """
+    pattern = re.compile("name: '?\"?" + name + "'?\"?$", re.MULTILINE)
```
Comment (spiffxp) [inline experiment/migrate_testgrid_tabs.py]: nit: you are matching more than just prow job names with this (env var names, volume names, etc)

This is getting lucky that none of those match a prowjob's name

**Comment by spiffxp** [comment]
/lgtm cancel
actually, wait, one commit message hygiene thing, please drop reference to #13057 in the commit message

**Comment by spiffxp** [comment]
(this is why I'm nitting on the commit message)
Screen Shot 2019-07-08 at 3 19 24 PM

**Pull request merged**